--- basic.py	2016-08-05 12:30:34.343165000 +0900
+++ basic.new.py	2016-08-05 12:26:47.792439900 +0900
@@ -1808,6 +1808,71 @@
     def backup_local(self, fn):
         '''make a date-marked backup of the specified file, return True or False on success or failure'''
 
+##########################################################################
+        cur_date = time.strftime("%Y%m%d", time.localtime(time.time()))
+        cur_time = time.strftime("%H%M%S", time.localtime(time.time()))
+
+        # backup settings
+        backup_generation = 5
+        backdest_root = '/var/log/ansible/backup'
+        backdest_prefix = os.path.join(backdest_root, cur_date)
+
+        # remove old backup
+        bk_dir_list = os.listdir(backdest_root)
+        bk_dir_list.sort()
+        while len(bk_dir_list) > backup_generation:
+            bk_dir_name = bk_dir_list.pop(0)
+            del_dir = os.path.join(backdest_root, bk_dir_name)
+            if os.path.isdir(del_dir):
+                shutil.rmtree(del_dir)
+            else:
+                os.remove(del_dir)
+
+        # archive old backup
+        import tarfile
+        bk_dir_list2 = os.listdir(backdest_root)
+        while len(bk_dir_list2) > 0:
+            bk_dir_name = bk_dir_list2.pop()
+            archive_dir = os.path.join(backdest_root, bk_dir_name)
+
+            if (bk_dir_name == cur_date):
+                continue
+
+            if os.path.isdir(archive_dir):
+                old_cwd = os.getcwd()
+                os.chdir(backdest_root)
+
+                archive_file = '%s.tar.gz' % bk_dir_name
+                n = 0
+                while os.path.exists(archive_file):
+                    archive_file = '%s_%d.tar.gz' % (bk_dir_name, n)
+                    n = n + 1
+
+                tf = tarfile.open(archive_file, 'w:gz')
+                for _root, _dirs, _files in os.walk(bk_dir_name):
+                    for _file in _files:
+                        tf.add(os.path.join(_root, _file))
+                tf.close()
+
+                shutil.rmtree(bk_dir_name)
+                os.chdir(old_cwd)
+
+        # ext_format: '<filename>.<date>-<time>-<pid>~'
+        ext = '%s-%s-%s~' % (cur_date, cur_time, os.getpid())
+        backupdest = '%s%s.%s' % (backdest_prefix, fn, ext)
+
+        if os.path.exists(fn):
+            try:
+                if not os.path.exists(os.path.dirname(backupdest)):
+                    os.makedirs(os.path.dirname(backupdest))
+                shutil.copy2(fn, backupdest)
+            except (shutil.Error, IOError):
+                e = get_exception()
+                self.fail_json(msg='Could not make backup of %s to %s: %s' % (fn, backupdest, e))
+
+        return backupdest
+##########################################################################
+
         backupdest = ''
         if os.path.exists(fn):
             # backups named basename-YYYY-MM-DD@HH:MM:SS~
